### PR TITLE
Dependencies: refactor API dependencies, fix proto JSON block, finish pip install hashes

### PR DESCRIPTION
**Pull request by moderation**
Signed-off-by: moderation michael@sooper.org

Risk Level: Low
Testing:
- `bazel --nohome_rc build --jobs 12 --ui_actions_shown 12 @envoy_api_canonical//envoy/...`
- `docs/build.sh`

Docs Changes: none required
Release Notes: none required

Comment: Will enable addition of API dependencies in follow up to https://github.com/envoyproxy/envoy/pull/13340 /cc @htuch

### CODE REVIEW

**Review by htuch — APPROVED** [review]
Really nice, excited to have all proxy binary deps now with metadata.

**Comment by moderation** [comment]
@htuch I'm a little stuck as bumping the Pygments ([2.6.0 to 2.7.1](https://github.com/pygments/pygments/pull/1528)) Python dependency introduces more explicit JSON escaping. So docs fail on the following JSON snippets in `stats.proto`
```
    // .. code-block:: json
    //   {
    //     "tag_name": "envoy.cluster_name",
    //     "regex": "^cluster\.((.+?)\.)"
    //   }
    // .. code-block:: json
    //   [
    //     {
    //       "tag_name": "envoy.http_user_agent",
    //       "regex": "^http(?=\.).*?\.user_agent\.((.+?)\.)\w+?$"
    //     },
    //     {
    //       "tag_name": "envoy.http_conn_manager_prefix",
    //       "regex": "^http\.((.*?)\.)"
    //     }
    //   ]
```
But when fixing the snippets with `\\` vs `\` `Run check format scripts` breaks - https://dev.azure.com/cncf/envoy/_build/results?buildId=52723&view=logs&j=9fd74226-df5b-5f29-39da-76fc341cc763&t=068b98b1-acf0-52d0-fc6e-f6d1a5b1d5ef. I'm not able to repro locally with `env CLANG_FORMAT=/usr/local/bin/clang-format tools/code_format/check_format.py check` so still digging

**Comment by htuch** [comment]
@moderation is this just for Sphinx generation? In any case, I recommend splitting this Pygments upgrade into a separate PR if it's going to involve a lot of reasoning.

**Comment by moderation** [comment]
I think I've worked this out and could produce locally. Trying one me time

**Comment by moderation** [comment]
This is ready for final review @htuch + @envoyproxy/api-shepherds

File: api/bazel/repository_locations.bzl [hunk 1]
```diff
@@ -1,66 +1,110 @@
...
+        use_category = ["api"],
     ),
 )
+
+def _format_version(s, version):
+    return s.format(version = version, dash_version = version.replace(".", "-"), underscore_version = version.replace(".", "_"))
+
+# Interpolate {version} in the above dependency specs. This code should be capable of running in both Python
+# and Starlark.
+def _dependency_repositories():
```
Comment (htuch) [inline api/bazel/repository_locations.bzl:97]: FYI I will refactor this with bazel/repository_locations.bzl in a followup where APIs are added to the dashboard.

File: api/envoy/config/metrics/v2/stats.proto [hunk 2]
```diff
@@ -201,7 +201,7 @@ message TagSpecifier {
     //
     //   {
     //     "tag_name": "envoy.cluster_name",
-    //     "regex": "^cluster\.((.+?)\.)"
+    //     "regex": "^cluster\\.((.+?)\\.)"
```
Comment (htuch) [inline api/envoy/config/metrics/v2/stats.proto:204]: Is this a docs only change? Will the docs be misleading vs. config users need to plugin?

**Review by htuch — COMMENTED** [review]
LGTM but a question on the escaping changes.

File: api/envoy/config/metrics/v2/stats.proto (hunk 2 above)
Comment (moderation) [reply api/envoy/config/metrics/v2/stats.proto:204]: This should be a doc only change. The increased escaping enforcement in Pygment caught the issue with the existing code. What is there is not valid JSON. This is the only error produced in a full docs run with the new Python deps
image
image

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Comment by alyssawilk** [comment]
Hey @moderation thanks for making deps more clear!
I think this may have broken the deprecation versions script though - sh tools/deprecate_version/deprecate_version.sh now complains about having non-pinned deps.  Mind taking a look?

**Comment by moderation** [comment]
Sorry for this. Somehow this script isn't triggered in CI. I'll raise a PR ASAP to fix

```python
Deprecated==1.2.10 \
    --hash=sha256:525ba66fb5f90b07169fdd48b6373c18f1ee12728ca277ca44567a367d9d7f74 \
    --hash=sha256:a766c1dccb30c5f6eb2b203f87edd1d8588847709c78589e1521d769addc8218
gitdb==4.0.5 \
    --hash=sha256:91f36bfb1ab7949b3b40e23736db18231bf7593edada2ba5c3a174a7b23657ac \
    --hash=sha256:c9e1f2d0db7ddb9a704c2a0217be31214e91a4fe1dea1efad19ae42ba0c285c9
GitPython==3.1.8 \
    --hash=sha256:080bf8e2cf1a2b907634761c2eaefbe83b69930c94c66ad11b65a8252959f912 \
    --hash=sha256:1858f4fd089abe92ae465f01d5aaaf55e937eca565fb2c1fce35a51b5f85c910
PyGithub==1.53 \
    --hash=sha256:776befaddab9d8fddd525d52a6ca1ac228cf62b5b1e271836d766f4925e1452e \
    --hash=sha256:8ad656bf79958e775ec59f7f5a3dbcbadac12147ae3dc42708b951064096af15
smmap==3.0.4 \
    --hash=sha256:54c44c197c819d5ef1991799a7e30b662d1e520f2ac75c9efbeb54a742214cf4 \
    --hash=sha256:9c98bbd1f9786d22f14b3d4126894d56befb835ec90cef151af566c7e19b5d24
wrapt==1.12.1 \
    --hash=sha256:b62ffa81fb85f4332a4f609cab4ac40709470da05643a082ec1eb88e6d9b97d7
```

**Comment by alyssawilk** [comment]
No worries, as you say it doesn't run in CI because of the credentials issue (we run manually when we cut the release).  I suppose we could have an early exit version which didn't require the credentials just to catch things like this but it's generally not a biggie to just fix.

**Pull request merged**